make/StaticLink.gmk

```diff
@@ -0,0 +1,89 @@
+# Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+#
+# Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+# or visit www.oracle.com if you need additional information or have any
+# questions.
+#
+
+default: all
+
+include $(SPEC)
+include $(TOPDIR)/make/common/MakeBase.gmk
+include $(TOPDIR)/make/common/Modules.gmk
+
+$(eval $(call IncludeCustomExtension, StaticLink.gmk))
+
+################################################################################
+
+JAVASTATIC := javastatic$(EXE_SUFFIX)
+
+ORIGIN_ARG := $(call SET_EXECUTABLE_ORIGIN,/.)
+
+JAVASTATIC_OUTPUT_DIR := $(JDK_OUTPUTDIR)/bin
+JAVASTATIC_OBJECT_DIR := $(SUPPORT_OUTPUTDIR)/native/java.base/java
+STATIC_LIBS_DIR := $(STATIC_LIBS_IMAGE_DIR)/lib
+
+JDK_STATIC_LIBS := $(call FindFiles, $(STATIC_LIBS_IMAGE_DIR)/lib)
+# Currently we are supporting headless only.
+EXCLUDED_STATIC_LIBS := \
+  $(STATIC_LIBS_IMAGE_DIR)/lib/$(LIBRARY_PREFIX)jsound$(STATIC_LIBRARY_SUFFIX) \
+  $(STATIC_LIBS_IMAGE_DIR)/lib/$(LIBRARY_PREFIX)saproc$(STATIC_LIBRARY_SUFFIX) \
+  $(STATIC_LIBS_IMAGE_DIR)/lib/$(LIBRARY_PREFIX)jawt$(STATIC_LIBRARY_SUFFIX) \
+  $(STATIC_LIBS_IMAGE_DIR)/lib/$(LIBRARY_PREFIX)awt_xawt$(STATIC_LIBRARY_SUFFIX) \
+  $(STATIC_LIBS_IMAGE_DIR)/lib/$(LIBRARY_PREFIX)splashscreen$(STATIC_LIBRARY_SUFFIX) \
+  $(STATIC_LIBS_IMAGE_DIR)/lib/$(LIBRARY_PREFIX)jpackageapplauncher$(STATIC_LIBRARY_SUFFIX) \
+  $(STATIC_LIBS_IMAGE_DIR)/lib/$(LIBRARY_PREFIX)jpackageapplauncheraux$(STATIC_LIBRARY_SUFFIX)
+
+JVM_STATIC_LIB := $(STATIC_LIBS_IMAGE_DIR)/lib/$(LIBRARY_PREFIX)jvm$(STATIC_LIBRARY_SUFFIX)
+
+JAVASTATIC_LINK_LIBS := $(filter-out $(EXCLUDED_STATIC_LIBS), $(JDK_STATIC_LIBS)) $(JVM_STATIC_LIB)
+
+JAVASTATIC_LOG_PREFIX := $(JAVASTATIC_OBJECT_DIR)/BUILD_LAUNCHER_javastatic$(EXE_SUFFIX)_static_link
+
+JAVASTATIC_TARGET := $(JAVASTATIC_OUTPUT_DIR)/$(JAVASTATIC)
+JAVASTATIC_TARGET_IMAGE := $(JDK_IMAGE_DIR)/bin/$(JAVASTATIC)
+
+JAVASTATIC_LDFLAGS := $(JVM_LDFLAGS) $(ORIGIN_ARG) -Wl$(COMMA)--export-dynamic
+JAVASTATIC_LD_OBJ_ARG := $(JAVASTATIC_OBJECT_DIR)/main.o
+
+# Link with headless native only for now.
+JAVASTATIC_LIBS := -Wl$(COMMA)--whole-archive $(JAVASTATIC_LINK_LIBS)
+JAVASTATIC_EXTRA_LIBS := -Wl$(COMMA)--no-whole-archive -lpthread $(LIBDL) $(LIBM) $(LDFLAGS_CXX_STATIC_JDK)
+
+$(JVM_STATIC_LIB) : $(HOTSPOT_OUTPUTDIR)/variant-server/libjvm/objs/static/$(LIBRARY_PREFIX)jvm$(STATIC_LIBRARY_SUFFIX)
+	$(CP) $< $@
+
+$(JAVASTATIC_TARGET_IMAGE): $(JAVASTATIC_LINK_LIBS)
+        ifeq ($(call isTargetOs, linux), true)
+	  $(ECHO) "Creating '$(JAVASTATIC_TARGET_IMAGE)'"
+	  $(call ExecuteWithLog, $(JAVASTATIC_LOG_PREFIX), \
+	      $(LD) $(JAVASTATIC_LDFLAGS) $(LD_OUT_OPTION)$(JAVASTATIC_TARGET) \
+	          $(JAVASTATIC_LD_OBJ_ARG) $(JAVASTATIC_LIBS) $(JAVASTATIC_EXTRA_LIBS))
+	  $(CP) -f $(JAVASTATIC_TARGET) $@
+        else
+	  $(error Not supported on non-linux platform currently)
+        endif
```

Comment:
I think it would be rather beneficial if this linking step could be performed by some variant of SetupNativeCompilation. We don't want to have to maintain logic for dealing with different platforms and linkers outside of the common files if it can be helped. SetupNativeCompilation would need to be tweaked to run without SRC and just a set of `.a` files. 

The error about platform support should be handled in configure. We shouldn't be able to get here unless the configuration is expected to be supported.

Reply:
> I think it would be rather beneficial if this linking step could be performed by some variant of SetupNativeCompilation. We don't want to have to maintain logic for dealing with different platforms and linkers outside of the common files if it can be helped. SetupNativeCompilation would need to be tweaked to run without SRC and just a set of .a files.
>
> The error about platform support should be handled in configure. We shouldn't be able to get here unless the configuration is expected to be supported.

All sounds good, thanks. I plan to withdraw this PR and extract the .a handling part into a new PR for now (as mentioned in response to @AlanBateman's comments). For the linking part, it would be good to have your close/direct involvement and collaboration in branch.